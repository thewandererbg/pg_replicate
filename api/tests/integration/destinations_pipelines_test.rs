use reqwest::StatusCode;

use crate::{
    common::test_app::{
        spawn_test_app, CreateDestinationPipelineResponse, CreatePipelineRequest,
        DestinationResponse, PipelineResponse, PostDestinationPipelineRequest,
    },
    integration::destination_test::{
        create_destination, new_destination_config, new_name, updated_destination_config,
        updated_name,
    },
    integration::images_test::create_default_image,
    integration::pipelines_test::{new_pipeline_config, updated_pipeline_config},
    integration::sources_test::create_source,
    integration::tenants_test::{create_tenant, create_tenant_with_id_and_name},
};

#[tokio::test(flavor = "multi_thread")]
async fn destination_and_pipeline_can_be_created() {
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;
    let source_id = create_source(&app, tenant_id).await;
    create_default_image(&app).await;

    // Act
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: new_name(),
        destination_config: new_destination_config(),
        source_id,
        pipeline_config: new_pipeline_config(),
    };
    let response = app
        .create_destination_pipeline(tenant_id, &destination_pipeline)
        .await;

    // Assert
    assert!(response.status().is_success());
    let response: CreateDestinationPipelineResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.destination_id, 1);
    assert_eq!(response.pipeline_id, 1);

    let destination_id = response.destination_id;
    let pipeline_id = response.pipeline_id;

    let response = app.read_destination(tenant_id, destination_id).await;
    let response: DestinationResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.id, destination_id);
    assert_eq!(response.name, destination_pipeline.destination_name);
    insta::assert_debug_snapshot!(response.config);

    let response = app.read_pipeline(tenant_id, pipeline_id).await;
    let response: PipelineResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.id, pipeline_id);
    assert_eq!(&response.tenant_id, tenant_id);
    assert_eq!(response.source_id, source_id);
    assert_eq!(response.destination_id, destination_id);
    assert_eq!(response.replicator_id, 1);
    insta::assert_debug_snapshot!(response.config);
}

#[tokio::test(flavor = "multi_thread")]
async fn destination_and_pipeline_with_another_tenants_source_cant_be_created() {
    // Arrange
    let app = spawn_test_app().await;
    create_default_image(&app).await;
    let tenant1_id = &create_tenant_with_id_and_name(
        &app,
        "abcdefghijklmnopqrst".to_string(),
        "tenant_1".to_string(),
    )
    .await;
    let tenant2_id = &create_tenant_with_id_and_name(
        &app,
        "tsrqponmlkjihgfedcba".to_string(),
        "tenant_2".to_string(),
    )
    .await;
    let source2_id = create_source(&app, tenant2_id).await;

    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: new_name(),
        destination_config: new_destination_config(),
        source_id: source2_id,
        pipeline_config: new_pipeline_config(),
    };
    let response = app
        .create_destination_pipeline(tenant1_id, &destination_pipeline)
        .await;

    // Assert
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
}

#[tokio::test(flavor = "multi_thread")]
async fn an_existing_destination_and_pipeline_can_be_updated() {
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;
    let source_id = create_source(&app, tenant_id).await;
    create_default_image(&app).await;
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: new_name(),
        destination_config: new_destination_config(),
        source_id,
        pipeline_config: new_pipeline_config(),
    };
    let response = app
        .create_destination_pipeline(tenant_id, &destination_pipeline)
        .await;
    let response: CreateDestinationPipelineResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let CreateDestinationPipelineResponse {
        destination_id,
        pipeline_id,
    } = response;
    let new_source_id = create_source(&app, tenant_id).await;

    // Act
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: updated_name(),
        destination_config: updated_destination_config(),
        source_id: new_source_id,
        pipeline_config: updated_pipeline_config(),
    };
    let response = app
        .update_destination_pipeline(
            tenant_id,
            destination_id,
            pipeline_id,
            &destination_pipeline,
        )
        .await;

    // Assert
    assert!(response.status().is_success());

    let response = app.read_destination(tenant_id, destination_id).await;
    let response: DestinationResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.id, destination_id);
    assert_eq!(response.name, destination_pipeline.destination_name);
    insta::assert_debug_snapshot!(response.config);

    let response = app.read_pipeline(tenant_id, pipeline_id).await;
    let response: PipelineResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.id, pipeline_id);
    assert_eq!(&response.tenant_id, tenant_id);
    assert_eq!(response.source_id, destination_pipeline.source_id);
    assert_eq!(response.destination_id, destination_id);
    assert_eq!(response.replicator_id, 1);
    insta::assert_debug_snapshot!(response.config);
}

#[tokio::test(flavor = "multi_thread")]
async fn destination_and_pipeline_with_another_tenants_source_cant_be_updated() {
    // Arrange
    let app = spawn_test_app().await;
    create_default_image(&app).await;
    let tenant1_id = &create_tenant_with_id_and_name(
        &app,
        "abcdefghijklmnopqrst".to_string(),
        "tenant_1".to_string(),
    )
    .await;
    let tenant2_id = &create_tenant_with_id_and_name(
        &app,
        "tsrqponmlkjihgfedcba".to_string(),
        "tenant_2".to_string(),
    )
    .await;

    let source1_id = create_source(&app, tenant1_id).await;
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: new_name(),
        destination_config: new_destination_config(),
        source_id: source1_id,
        pipeline_config: new_pipeline_config(),
    };
    let response = app
        .create_destination_pipeline(tenant1_id, &destination_pipeline)
        .await;
    let response: CreateDestinationPipelineResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let CreateDestinationPipelineResponse {
        destination_id,
        pipeline_id,
    } = response;

    // Act
    let source2_id = create_source(&app, tenant2_id).await;
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: updated_name(),
        destination_config: updated_destination_config(),
        source_id: source2_id,
        pipeline_config: updated_pipeline_config(),
    };
    let response = app
        .update_destination_pipeline(
            tenant1_id,
            destination_id,
            pipeline_id,
            &destination_pipeline,
        )
        .await;

    // Assert
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
}

#[tokio::test(flavor = "multi_thread")]
async fn destination_and_pipeline_with_another_tenants_destination_cant_be_updated() {
    // Arrange
    let app = spawn_test_app().await;
    create_default_image(&app).await;
    let tenant1_id = &create_tenant_with_id_and_name(
        &app,
        "abcdefghijklmnopqrst".to_string(),
        "tenant_1".to_string(),
    )
    .await;
    let tenant2_id = &create_tenant_with_id_and_name(
        &app,
        "tsrqponmlkjihgfedcba".to_string(),
        "tenant_2".to_string(),
    )
    .await;

    let source1_id = create_source(&app, tenant1_id).await;
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: new_name(),
        destination_config: new_destination_config(),
        source_id: source1_id,
        pipeline_config: new_pipeline_config(),
    };
    let response = app
        .create_destination_pipeline(tenant1_id, &destination_pipeline)
        .await;
    let response: CreateDestinationPipelineResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let CreateDestinationPipelineResponse { pipeline_id, .. } = response;

    // Act
    let destination2_id = create_destination(&app, tenant2_id).await;
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: updated_name(),
        destination_config: updated_destination_config(),
        source_id: source1_id,
        pipeline_config: updated_pipeline_config(),
    };
    let response = app
        .update_destination_pipeline(
            tenant1_id,
            destination2_id,
            pipeline_id,
            &destination_pipeline,
        )
        .await;

    // Assert
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
}

#[tokio::test(flavor = "multi_thread")]
async fn destination_and_pipeline_with_another_tenants_pipeline_cant_be_updated() {
    // Arrange
    let app = spawn_test_app().await;
    create_default_image(&app).await;
    let tenant1_id = &create_tenant_with_id_and_name(
        &app,
        "abcdefghijklmnopqrst".to_string(),
        "tenant_1".to_string(),
    )
    .await;
    let tenant2_id = &create_tenant_with_id_and_name(
        &app,
        "tsrqponmlkjihgfedcba".to_string(),
        "tenant_2".to_string(),
    )
    .await;

    let source1_id = create_source(&app, tenant1_id).await;
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: new_name(),
        destination_config: new_destination_config(),
        source_id: source1_id,
        pipeline_config: new_pipeline_config(),
    };
    let response = app
        .create_destination_pipeline(tenant1_id, &destination_pipeline)
        .await;
    let response: CreateDestinationPipelineResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let CreateDestinationPipelineResponse {
        destination_id: destination1_id,
        ..
    } = response;

    let source2_id = create_source(&app, tenant2_id).await;
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: new_name(),
        destination_config: new_destination_config(),
        source_id: source2_id,
        pipeline_config: new_pipeline_config(),
    };
    let response = app
        .create_destination_pipeline(tenant2_id, &destination_pipeline)
        .await;
    let response: CreateDestinationPipelineResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let CreateDestinationPipelineResponse {
        pipeline_id: pipeline2_id,
        ..
    } = response;

    // Act
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: updated_name(),
        destination_config: updated_destination_config(),
        source_id: source1_id,
        pipeline_config: updated_pipeline_config(),
    };
    let response = app
        .update_destination_pipeline(
            tenant1_id,
            destination1_id,
            pipeline2_id,
            &destination_pipeline,
        )
        .await;

    // Assert
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
}

#[tokio::test(flavor = "multi_thread")]
async fn duplicate_destination_pipeline_with_same_source_cant_be_created() {
    // Arrange
    let app = spawn_test_app().await;
    create_default_image(&app).await;
    let tenant_id = &create_tenant(&app).await;
    let source_id = create_source(&app, tenant_id).await;

    // Create first destination and pipeline
    let destination_pipeline = PostDestinationPipelineRequest {
        destination_name: new_name(),
        destination_config: new_destination_config(),
        source_id,
        pipeline_config: new_pipeline_config(),
    };
    let response = app
        .create_destination_pipeline(tenant_id, &destination_pipeline)
        .await;
    assert!(response.status().is_success());
    let response: CreateDestinationPipelineResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let first_destination_id = response.destination_id;

    // Act - Try to create another pipeline with same source and the first destination
    let pipeline_request = CreatePipelineRequest {
        source_id,
        destination_id: first_destination_id,
        config: updated_pipeline_config(),
    };
    let response = app.create_pipeline(tenant_id, &pipeline_request).await;

    // Assert
    assert_eq!(response.status(), StatusCode::CONFLICT);
}
