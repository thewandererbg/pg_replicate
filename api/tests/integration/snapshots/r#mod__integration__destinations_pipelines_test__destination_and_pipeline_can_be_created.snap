---
source: api/tests/integration/destinations_pipelines_test.rs
expression: response.config
---
BigQuery {
    project_id: "project-id",
    dataset_id: "dataset-id",
    service_account_key: "REDACTED",
    max_staleness_mins: None,
}
